gno.land/cmd/gnoland/testdata/event_defer_callback_loop.txtar

```diff
@@ -0,0 +1,68 @@
+# start a new node
+gnoland start
+
+gnokey maketx addpkg -pkgdir $WORK -pkgpath gno.land/r/demo/edcl -gas-fee 1000000ugnot -gas-wanted 2500000 -broadcast -chainid=tendermint_test test1
+stdout OK!
+stdout 'GAS WANTED: 2500000'
+stdout 'GAS USED:   [0-9]+'
+stdout 'HEIGHT:     [0-9]+'
+stdout 'EVENTS:     \[\]'
+
+gnokey maketx call -pkgpath gno.land/r/demo/edcl -func Main -gas-fee 1000000ugnot -gas-wanted 2000000 -broadcast -chainid=tendermint_test test1
+stdout OK!
+stdout 'GAS WANTED: 2000000'
+stdout 'GAS USED:   [0-9]+'
+stdout 'HEIGHT:     [0-9]+'
+stdout 'EVENTS:     \[{\"type\":\"ForLoopEvent\",\"pkg_path\":\"gno.land\/r\/demo\/edcl\",\"func\":\"\",\"attributes\":\[{\"key\":\"iteration\",\"value\":\"0\"},{\"key\":\"key\",\"value\":\"value\"}\]},{\"type\":\"ForLoopEvent\",\"pkg_path\":\"gno.land\/r\/demo\/edcl\",\"func\":\"\",\"attributes\":\[{\"key\":\"iteration\",\"value\":\"1\"},{\"key\":\"key\",\"value\":\"value\"}\]},{\"type\":\"ForLoopEvent\",\"pkg_path\":\"gno.land\/r\/demo\/edcl\",\"func\":\"\",\"attributes\":\[{\"key\":\"iteration\",\"value\":\"2\"},{\"key\":\"key\",\"value\":\"value\"}\]},{\"type\":\"ForLoopCompletionEvent\",\"pkg_path\":\"gno.land\/r\/demo\/edcl\",\"func\":\"forLoopEmitExample\",\"attributes\":\[{\"key\":\"count\",\"value\":\"3\"}\]},{\"type\":\"CallbackEvent\",\"pkg_path\":\"gno.land\/r\/demo\/edcl\",\"func\":\"\",\"attributes\":\[{\"key\":\"key1\",\"value\":\"value1\"},{\"key\":\"key2\",\"value\":\"value2\"}\]},{\"type\":\"CallbackCompletionEvent\",\"pkg_path\":\"gno.land\/r\/demo\/edcl\",\"func\":\"callbackEmitExample\",\"attributes\":\[{\"key\":\"key\",\"value\":\"value\"}\]},{\"type\":\"DeferEvent\",\"pkg_path\":\"gno.land\/r\/demo\/edcl\",\"func\":\"deferEmitExample\",\"attributes\":\[{\"key\":\"key1\",\"value\":\"value1\"},{\"key\":\"key2\",\"value\":\"value2\"}\]}\]'
```

Comment:
```
{\"type\":\"ForLoopEvent\",\"pkg_path\":\"gno.land\/r\/demo\/edcl\",\"func\":\"\",\"attributes\":\
```

I've identified three issues in the current JSON output:
1. Changing `attributes` to `attrs` can save bytes while remaining human-readable.
2. The `func` field is empty, and should not, I think we should take the most recent "named" frame.
3. Consider adding a default field `prev_realm` to indicate the caller. This helps differentiate it from `pkg_path` and `func`, which pertain to the code defining the `std.Emit` implementation (great). The purpose is similar to assigning assets and bankers to prevRealm instead of origCall or the called contract in synchronous calls.

Reply:
I'm okay with someone fixing 2 and 3 in another PR if there's an issue to ensure we don't forget.

Reply:
We should also mention in the same issue the need to write documentation explaining the new feature and providing usage examples. Please cc @gnolang/devrels.

Reply:
thanks! the first one has resolved in https://github.com/gnolang/gno/pull/1653/commits/7a5fd741d4b2fbd5857b8055abb6c08c8c11e4be

In the second case, where the function name is empty, it seems that the function name cannot be retrieved from the stack frame only for special purpose functions such as anonymous functions (or lambda functions) used in callbacks or defer, or test functions.

For the last third case, it seems that it can be handled by creating a getPrevRealm function, similar to the function name searcher. It can be done in a way like this: 

```go
// Pseudo code for previous realm name getter
func getPrevRealm(m *gno.Machine, height int) (address string, pkgPath string) {
    var (
        ctx           = m.Context.(ExecContext)
        currentCaller crypto.Bech32Address
        // Keeps track of the number of times currentCaller has changed.
        changes int
    )

    for i := m.NumFrames() - 1; i > 0; i-- {
        fr := m.Frames[i]
        if fr.LastPackage == nil || !fr.LastPackage.IsRealm() {
            continue
        }

        // LastPackage is a realm. Get caller and pkgPath, and compare against
        // current* values.
        caller := fr.LastPackage.GetPkgAddr().Bech32()
        pkgPath := fr.LastPackage.PkgPath
        if caller != currentCaller {
            currentCaller = caller
            changes++
            if changes == height + 1 {
                return string(caller), pkgPath
            }
        }
    }

    // Fallback case: return OrigCaller.
    return string(ctx.OrigCaller), ""
}
```
receiving `height` as an argument and returning the `(height + 1)`th change. I think it would be good to gather people's opinions and include the necessary data.

Reply:
I've opened up issues for 2. and 3:

- https://github.com/gnolang/gno/issues/2001
- https://github.com/gnolang/gno/issues/2002

And of course the documentation:
- https://github.com/gnolang/gno/issues/2003

Reply:
> For the last third case, it seems that it can be handled by creating a getPrevRealm function, similar to the function name searcher. It can be done in a way like this:

I believe we now have a helper, likely introduced in #667 or a subsequent PR.